$$f'(2) - ?$$

In [1]:
import math
from sympy.matrices import Matrix
import sympy
from sympy.abc import x,y
from sympy import *
init_printing( use_latex='mathjax' )
f = Function("f")

In [2]:
f=[1, 0.5, 0.3, 0.2, 0.1]
val=[0, 1, 2, 5, 7]

Строим интерполяционный полином Лагранжа

In [3]:
L_1=Function("L_1")
L_1=(x-val[1])*(x-val[2])*(x-val[3])*(x-val[4])/((val[0]-val[1])*(val[0]-val[2])*(val[0]-val[3])*(val[0]-val[4]))
L_1

(x - 7)⋅(x - 5)⋅(x - 2)⋅(x - 1)
───────────────────────────────
               70              

In [4]:
L_2=Function("L_2")
L_2=(x-val[0])*(x-val[2])*(x-val[3])*(x-val[4])/((val[1]-val[0])*(val[1]-val[2])*(val[1]-val[3])*(val[1]-val[4]))
L_2

-x⋅(x - 7)⋅(x - 5)⋅(x - 2) 
───────────────────────────
             24            

In [5]:
L_3=Function("L_3")
L_3=(x-val[0])*(x-val[1])*(x-val[3])*(x-val[4])/((val[2]-val[0])*(val[2]-val[1])*(val[2]-val[3])*(val[2]-val[4]))
L_3

x⋅(x - 7)⋅(x - 5)⋅(x - 1)
─────────────────────────
            30           

In [6]:
L_4=Function("L_4")
L_4=(x-val[0])*(x-val[1])*(x-val[2])*(x-val[4])/((val[3]-val[0])*(val[3]-val[1])*(val[3]-val[2])*(val[3]-val[4]))
L_4

-x⋅(x - 7)⋅(x - 2)⋅(x - 1) 
───────────────────────────
            120            

In [7]:
L_5=Function("L_5")
L_5=(x-val[0])*(x-val[1])*(x-val[2])*(x-val[3])/((val[4]-val[0])*(val[4]-val[1])*(val[4]-val[2])*(val[4]-val[3]))
L_5

x⋅(x - 5)⋅(x - 2)⋅(x - 1)
─────────────────────────
           420           

Итоговый многочлен равен

In [8]:
L=Function("L")
L=f[0]*L_1+f[1]*L_2+f[2]*L_3+f[3]*L_4+f[4]*L_5
L

                                                                              
-0.0208333333333333⋅x⋅(x - 7)⋅(x - 5)⋅(x - 2) + 0.01⋅x⋅(x - 7)⋅(x - 5)⋅(x - 1)
                                                                              

                                                                              
 - 0.00166666666666667⋅x⋅(x - 7)⋅(x - 2)⋅(x - 1) + 0.000238095238095238⋅x⋅(x -
                                                                              

                      (x - 7)⋅(x - 5)⋅(x - 2)⋅(x - 1)
 5)⋅(x - 2)⋅(x - 1) + ───────────────────────────────
                                     70              

In [9]:
g = diff(L)
g

                                                                              
-0.0108333333333333⋅x⋅(x - 7)⋅(x - 5) - 0.0225⋅x⋅(x - 7)⋅(x - 2) + 0.008333333
                                                                              

                                                                              
33333333⋅x⋅(x - 7)⋅(x - 1) - 0.0205952380952381⋅x⋅(x - 5)⋅(x - 2) + 0.01023809
                                                                              

                                                                              
52380952⋅x⋅(x - 5)⋅(x - 1) - 0.00142857142857143⋅x⋅(x - 2)⋅(x - 1) + (0.001666
                                                                              

                                                                              
66666666667 - 0.00166666666666667⋅x)⋅(x - 7)⋅(x - 2) + (0.145833333333333 - 0.
                                                                              

                                                

In [10]:
g.subs(x,2)

-0.105476190476190

Методом неопределенных коэффициентов

Подберем коэффициенты $a,b,c,d,e$ так, что бы 
\begin{equation}
af(x_{0})+bf(x_{1})+cf(x_{2})+df(x_{3})+ef(x_{4}) = f'(x_2)
\end{equation}

Разложив в ряд Тейлора получим:

\begin{equation}
a \left(f(x_{2}) + (x_{0}-x_{2})f'(x_{2}) + \frac{(x_{0}-x_{2})^2}{2}f''(x_{2}) + \frac{(x_{0}-x_{2})^3}{6}f'''(x_{2}) + \frac{(x_{0}-x_{2})^4}{24}f^{IV}(x_{2}) + \frac{(x_{0}-x_{2})^5}{120}f^{V}(\theta) \right) +
b \left(f(x_{2}) + (x_{1}-x_{2})f'(x_{2}) + \frac{(x_{1}-x_{2})^2}{2}f''(x_{2}) + \frac{(x_{1}-x_{2})^3}{6}f'''(x_{2}) + \frac{(x_{1}-x_{2})^4}{24}f^{IV}(x_{2}) + \frac{(x_{1}-x_{2})^5}{120}f^{V}(\theta) \right) +
c \cdot f(x_{2}) +
d \left(f(x_{2}) + (x_{3}-x_{2})f'(x_{2}) + \frac{(x_{3}-x_{2})^2}{2}f''(x_{2}) + \frac{(x_{3}-x_{2})^3}{6}f'''(x_{2}) + \frac{(x_{3}-x_{2})^4}{24}f^{IV}(x_{2}) + \frac{(x_{3}-x_{2})^5}{120}f^{V}(\theta) \right) +
e \left(f(x_{2}) + (x_{4}-x_{2})f'(x_{2}) + \frac{(x_{4}-x_{2})^2}{2}f''(x_{2}) + \frac{(x_{4}-x_{2})^3}{6}f'''(x_{2}) + \frac{(x_{4}-x_{2})^4}{24}f^{IV}(x_{2}) + \frac{(x_{4}-x_{2})^5}{120}f^{V}(\theta) \right) = f'(x_2)
\end{equation}

Откуда и систему на коэфициенты:

\begin{cases}
   a + b + c + d + e = 0\\
   -2a -b +3d + 5e = 1\\
   4a + b + 9d + 25e = 0\\
   -8a - b + 27d + 125e = 0\\
   16a + b + 81d + 625e = 0
\end{cases}

In [11]:
M=Matrix([[1,1,1,1,1], [-2,-1,0,3,5], [4,1,0,9,25], [-8,-1,0,27,125], [16,1,0,81,625]])
M

⎡1   1   1  1    1 ⎤
⎢                  ⎥
⎢-2  -1  0  3    5 ⎥
⎢                  ⎥
⎢4   1   0  9   25 ⎥
⎢                  ⎥
⎢-8  -1  0  27  125⎥
⎢                  ⎥
⎣16  1   0  81  625⎦

In [12]:
b = Matrix([0,1,0,0,0])
b

⎡0⎤
⎢ ⎥
⎢1⎥
⎢ ⎥
⎢0⎥
⎢ ⎥
⎢0⎥
⎢ ⎥
⎣0⎦

In [13]:
sol=M.inv()*b
sol

⎡3/14 ⎤
⎢     ⎥
⎢-5/4 ⎥
⎢     ⎥
⎢ 29  ⎥
⎢ ──  ⎥
⎢ 30  ⎥
⎢     ⎥
⎢1/12 ⎥
⎢     ⎥
⎣-1/70⎦

In [14]:
func=Matrix([f[0],f[1],f[2],f[3],f[4]])
func.T

[1  0.5  0.3  0.2  0.1]

In [15]:
func.T*sol

[-0.10547619047619]